Rudimentary support for contact and calendar

jaudriga opened the pull request:
Adds rudimentary support for contacts and calendars as it is currently specified at https://jmap.io/ .

More specifically this adds `/get` and `/set` methods for the `Contact`, `Calendar` and `CalendarEvent` entities. This PR's intention is to provide a basis for implementing further details of the contacts and calendar specs. I am not sure if this is all that you need to fix #19 .

Verified it works by executing the unit tests and using it with Fastmail and Cyrus version 3.2.0-rc1.

NOTE: It should be compatible with the JMAP features as implemented in https://github.com/cyrusimap/cyrus-imapd/tree/jmap-calendars-01 . However, we will verify by testing against that branch soon. I recommend waiting with the merge until then.

jaudriga commented:
We verified that the implemented method calls work against Cyrus built from above's feature branch.

However, the underlying objects differ slightly:
* `Calendar` objects have around 10-15 new properties and the "rights" properties have been refactored
* `CalendarEvent` have a few additional properties (including additional Fastmail-specific properties) and `recurrenceRule` is now an array

In its current form the PR implements what we saw in the stable version of Cyrus and the current Fastmail API. Changing the objects to match the feature branch would implement it for upcoming versions of Cyrus/Fastmail. I propse implementing it for the current versions first and opening up another PR for future versions. But of course I would be glad to hear your preferences on that.

jaudriga commented:
Pushed a new version with the following fixes:
* Use classes from Java 8 instead of `Date`
* Import newer jUnit packages
* Implemented a `DurationTypeAdapter` for the `java.time.Duration` type to fix JSON (de)serialization issues

iNPUTmice commented:
Hi,

thank you for bearing with me so far.

There are two things remaining:

* I don’t like the how CalanderTriggers are handled right now.
  * I don’t like that it uses SuperBuilders. I found SuperBuilders to be a bit nasty in some scenarios. (Requiring all classes to be annotated even if we don’t need builders for them, …)
  * You can set the type with the builder even though it should be hard coded per class. Currently I could set an ABSOLUTE_TRIGGER type when building an OffsetTrigger
  * Using `@Builder` on the constructor, which then doesn’t have the type argument can fix this.
  * Do CalendarTriggerType and CalendarType need to be the same enum?
  * Is the serializer necessary? From a quick glance it looks like an CalendarType would just serialize as is to the same JSON.  
* "PatchObject is essentially a CalendarEvent object, simply not with all properties set" is unfortunately not true. First of all PatchObjects can contain paths to nested objects. Second of all null is handled differently than in the rest of this library. In a normal entity like CalendarEvent null means missing. In a PatchObject null means please remove. So if you serialize a PatchObject to an entity (ignoring the path problem for a second) you wouldn’t know if a property is simple not covered by the patch (null) or should be removed by the patch (also null.)

This library currently doesn’t have a good API for patch objects as those a bit difficult to handle in any type safe manner. (I'm currently just using Map<String,Object>) and I’m open for better suggestions. But the approach in this PR is certainly wrong. (At least wronger that what the rest of the library is doing.)

jaudriga commented:
Thanks for reviewing it in more detail. We are working on fixing above's points.

I think resorting to using `Map<String,Object>` is the way to go here to keep the changes of this PR as small as necessary.

jaudriga commented:
Pushed a new version where we:
* removed SuperBuilders
* keep `CalendarTriggerType` and `CalendarType` in the same enum. `CalendarType` then contains all the `@type` properties that are defined in JSCalendar
* removed the serializer for `CalendarType`
* use `Map<String,Object>` for the PatchObject.

Some notes:
* We had used an own serializer for `CalendarAlertTrigger`, because without it we were unable to (de)serialize the object properly. However, now it serializes correctly, yes.

jaudriga commented:
Due to our recent discussion over XMPP, I now pushed a new version with the following changes:
* Split `ContactInformation` class into three subclasses. According to the [JMAP for Contacts spec](https://jmap.io/spec-contacts.html#contacts) the valid types for `ContactInformation` depends on the context. This change was necessary because previously it was possible to, e.g. assign an email address of type `HOME`.
* Turn JSCalendar's `@type` into an annotation and also add a serializer for JSCalendar objects.
* Add GSON adapter for LocalDateTime. Without it, `java.lang.NullPointerException` is thrown in some cases.

I chose to use a common interface `JSCalendar` for all classes that implement JSCalendar objects. While this is not perfect, I think this makes sense to have a common JSCalendar interface, because I see this as a first step to distinguish between JSCalendar format and JMAP for Calendars in the code. In my opinion this is a good idea, because:
1. The JSCalendar standard is close to being accepted. No major things will change here.
1. JSCalendar defines a common format that is usable beyond software that supports JMAP. E.g. in the future one might think about extracting JSCalendar files from the stored contact or calendar data.

iNPUTmice commented:
Hi,

Good news: I haven’t forgotten about this. I had been pretty busy with other projects in the last couple of month but now I’m back working on JMAP and Ltt.rs stuff.

I've just now implemented a - what I think is a pretty clean - solution for the `@type` problem.
The commit message has some rational on the why: https://github.com/iNPUTmice/jmap/commit/efddfac6c9e71f55ae9f2cb9b5830b3c1a29da4f ¹ but the general idea (and relevant for this PR) is that you can now simply annotate any entity with `@Type('aName')` and everything else will be taken care of.

For example you can now just use:
```java
@Getter
@Builder(toBuilder = true)
@Type("Alert")
public class CalendarAlert  {
	private CalendarTrigger trigger;

	private String acknowledged;

	private Map<String, CalendarRelation> relatedTo;

	private CalendarAlertAction action;

}
```

No need for the additional annotation processor (JSCalendarTypeProcessor) nor for the common base class `JSCalendar` or the custom adapter.

Note that in the above example if the class were to be called "Alert" one could simply omit value of `@Type` and simply write:
```java
@Type
public class Alert  {
    …
}
```

¹: I always had this solution in mind but I wasn’t entirely sure it would work before I wrote the code.

jaudriga commented:
Thanks. Looks good. I did not know you were planning to use it for other entities as well

I removed `JSCalendarTypeProcessor`, the `JSCalendar` class and the custom adapter.
